Documentation: How to contribute (#294)
SEKOIA-IO/intake-formats: doc/overview.md

```

#### XML

The `xml.parse-xml` stage can be used to transform a xml string into a dictionary.

> Per default, the `message` field is parsed but this property can be overwritten to specify any field.

Reviewer: ```suggestion
By default, the `message` field is parsed but this property can be overwritten to specify any field.
```

**Example**

In this example, the parser produces an event with fields that take their value from a deserialized xml field. Below is the example of the input event.

```json
{
  "message": "<Event><Id>1234</Id><Provider><Name>system</Name><Id>6</Id></Provider>"
}
```

> Below is the parsing pipeline that deserialize the `message` field and set the `agent.name` and `agent.id` fields.

Reviewer: ```suggestion
Below is the parsing pipeline that deserializes the `message` field and sets the `agent.name` and `agent.id` fields.
```

```yaml
pipeline:
  - name: parsed_event
    external:
      name: xml.parse-xml
  - name: agent
stages:
  agent:
    actions:
      - set:
          agent.id: '{{parsed_event.message.Event.Provider.Id}}'
          agent.name: '{{parsed_event.message.Event.Provider.Name}}'
```

The following shows the produced event.

```json
{
  "message": "<Event><Id>1234</Id><Provider><Name>system</Name><Id>6</Id></Provider>"
  "agent": {
    "id": 6,
    "name": "system"
  }
}
```

#### Windows

The `xml.parse-windows-event` stage can be used to transform a xml windows event into a dictionary.

> Per default, the `message` field is parsed but this property can be overwritten to specify any field.

Reviewer: ```suggestion
By default, the `message` field is parsed but this property can be overwritten to specify any field.
```

```json
{
  "message": "<Event><System><EventID>1234</EventID><Execution ProcessID=\"592\" ThreadID=\"6452\"/></System><EventData><Data Name=\"Key\">Value</Data></EventData></Event>"
}
```

Below is the parsing pipeline that deserialize the `message` field and set the fields.

```yaml
pipeline:
  - name: parsed_event
    external:
      name: xml.parse-windows-event
  - name: set-fields
stages:
  set-fields:
    actions:
      - set:
          process.pid: '{{parsed_event.message.System.Execution.ProcessID}}'
          process.thread.id: '{{parsed_event.message.System.Execution.ThreadID}}'
          event.id: '{{parsed_event.message.System.EventID}}'
          custom.key: '{{parsed_event.message.EventData.Key}}'
```

The following shows the produced event.

```json
{
  "message": "<Event><System><EventID>1234</EventID><Execution ProcessID=\"592\" ThreadID=\"6452\"/></System><EventData><Data Name=\"Key\">Value</Data></EventData></Event>",
  "process": {
    "pid": 592,
    "thread": {
      "id": 6452
    }
  },
  "event": {"id": 1234},
  "custom": {"key": "Value"}
}
```

#### CEF

The `cef.parse-cef` stage can be used to parse CEF messages.

> Per default, the `message` field is parsed but this property can be overwritten to specify any field.

Reviewer: ```suggestion
By default, the `message` field is parsed but this property can be overwritten to specify any field.
```

This stage will extract the following keys from the header definition:
- `CEFVersion`: The CEF version
- `DeviceVendor`: The vendor of the product that generated the log
- `DeviceProduct`: The product that generated the log
- `DeviceVersion`: The version of the product
- `DeviceEventClassID`: An unique identifier
- `Name`: A human-readable description of the event
- `Severity`: The severity of the event

and will extract each key-value from the extension.

**Example**

> In this example, the parser produces an event with fields that take their value from a CEF message. Below is the example of the input event.

Reviewer: ```suggestion
In this example, the parser produces an event with fields that take their value from a CEF message. Below is an example of the input event.
```

```json
{
  "message": "CEF:0|Security|threatmanager|1.0|100|worm successfully stopped|10|src=10.0.0.1 dst=2.1.2.2 spt=1232"
}
```

> Below is the parsing pipeline that deserialize the `message` field and set some fields.

Reviewer: ```suggestion
Below is the parsing pipeline that deserializes the `message` field and sets some fields.
```

  "source": {
    "ip": "10.0.0.1",
    "port": 1232
  },
  "destination": {
    "ip": "2.1.2.2"
  },
  "event": {"id": "100", "severity": 10},
  "observer": {
    "vendor": "Security",
    "product": "threatmanager",
    "version": "1.0"
  }
}
```

#### LEEF

The `leef.parse-leef` stage can be used to parse LEEF messages (supported version of LEEF are 1 and 2)
Per default, the `message` field is parsed but this property can be overwritten to specify any field.

This stage will extract the following keys from the header definition:
- `LEEFVersion`: The LEEF version
- `DeviceVendor`: The vendor of the product that generated the log
- `DeviceProduct`: The product that generated the log
- `DeviceVersion`: The version of the product
- `DeviceEventClassID`: An unique identifier

and will extract each key-value from the extension.

**Example**

In this example, the parser produces an event with fields that take their value from a CEF message. Below is the example of the input event.

```json
{
  "message": "LEEF:2|Security|threatmanager|1.0|100|^|src=10.0.0.1^dst=2.1.2.2^spt=1232"
}
```

> Below is the parsing pipeline that deserialize the `message` field and set some fields.

Reviewer: ```suggestion
Below is the parsing pipeline that deserializes the `message` field and sets some fields.
```

```yaml
pipeline:
  - name: parsed_event
    external:
      name: leef.parse-leef
  - name: agent
stages:
  agent:
    actions:
      - set:
          event.id: '{{parsed_event.message.DeviceEventClassID}}'
          source.ip: '{{parsed_event.message.src}}'
          source.port: '{{parsed_event.message.spt}}'
          destination.ip: '{{parsed_event.message.dst}}'
          observer.vendor: '{{parsed_event.message.DeviceVendor}}'
          observer.product: '{{parsed_event.message.DeviceProduct}}'
          observer.version: '{{parsed_event.message.DeviceVersion}}'
```

The following shows the produced event.

```json
{
  "message": "LEEF:2|Security|threatmanager|1.0|100|^|src=10.0.0.1^dst=2.1.2.2^spt=1232"
  "source": {
    "ip": "10.0.0.1",
    "port": 1232
  },
  "destination": {
    "ip": "2.1.2.2"
  },
  "event": {"id": "100"},
  "observer": {
    "vendor": "Security",
    "product": "threatmanager",
    "version": "1.0"
  }
}
```

## Action

> An action is an elementary operations that can create, update and delete fields.

Reviewer: ```suggestion
An action is an elementary operation that can create, update and delete fields.
```
